hal: Bulletproof kernel operations on user memory

etiaro opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes shortly -->
Inspired by BSD/openQNX exception handling routines, this draft introduces a mechanism with semantics similar to try - catch and based on idea similar to setjmp/longjmp by introducing 
- `hal_createexcjmp` that prepares "try-catch" context to be saved in TCB
- `threads_setexcjmp`/`threads_getexcjmp` which manages access to dedicated `thread_t` struct field

Introduced a few basic function and preprocessor macro for easy and safe use of this exception handling mechanism. 

On NOMMU targets, thanks to fixed configuration of accessible memory, simple access checks are good enough. That being said protection (R/W/E) checks are included in this PR to ensure no illegal accesses via kernel.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
There are some events in kernel code that require operating on user's memory., which is vulnerable to TOCTOU class races.
There's no good way to ensure such memory operations won't trigger a processor exception apart from interrupting all threads that could access that memory fragment.

These mechanisms can also be used in the future to gently handle user stackoverflows and faulty signal trampolines that currently can become a cascade of kernel exceptions.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

etiaro committed:
hal/armv7a: introduce CPU fault handling

Introduce usermem module with functions for safe operations on user
memory.

TASK: RTOS-1141

agkaminski commented on hal/armv7a/cpu.c:
```diff
@@ -110,19 +114,26 @@ int hal_cpuPushSignal(void *kstack, void (*handler)(void), cpu_context_t *signal
 	}
 	else {
-		signalCtx->psr &= ~THUMB_STATE;
+		threads_setexcjmp(oldctx, NULL);
```
Perhaps we could clear old ctx on `hal_excjmp`

agkaminski commented on proc/threads.c:
```diff
@@ -35,11 +36,15 @@ struct {
+
+	/* Synchronized by spinlock */
 	thread_t *ready[8];
+	/*
+	 * Write to current[i] only by i'th cpu, with spinlock set.
+	 * Read from current[i] by i'th cpu allowed without spinlock.
+	 */
```
Spinlock does nothing if it's used only for writes - let's just assume atomicity (and make these **pointers** volatile)

etiaro replied:
It's not used only for writes - it's used for reads on other cores.
The idea was that the i'th core it's not required because it cannot be writing at the same time as it's reading.

agkaminski commented on proc/threads.c:
```diff
@@ -1464,16 +1469,18 @@ static int _threads_checkSignal(thread_t *selected, process_t *proc, cpu_context
 		sig = hal_cpuGetLastBit(sig);
 
-		if (hal_cpuPushSignal(selected->kstack + selected->kstacksz, proc->sighandler, signalCtx, sig, oldmask, src) == 0) {
+		if ((ret = hal_cpuPushSignal(selected->kstack + selected->kstacksz, proc->sighandler, signalCtx, sig, oldmask, src)) == 0) {
```
```suggestion
		ret = hal_cpuPushSignal(selected->kstack + selected->kstacksz, proc->sighandler, signalCtx, sig, oldmask, src);
		if (ret == 0) {
```

agkaminski commented on proc/threads.c:
```diff
@@ -1491,18 +1498,29 @@ void threads_setupUserReturn(void *retval, cpu_context_t *ctx)
 
-	if (_threads_checkSignal(thread, thread->process, signalCtx, thread->sigmask, SIG_SRC_SCALL) == 0) {
+	// hal_spinlockClear(&threads_common.spinlock, &sc);  // TODO first load sigmask etc
+	int ret = _threads_checkSignal(thread, thread->process, signalCtx, thread->sigmask, SIG_SRC_SCALL);
```
Declare variables at the start (kernel, plo)

agkaminski commented on proc/threads.c:
```diff
@@ -1491,18 +1498,29 @@ void threads_setupUserReturn(void *retval, cpu_context_t *ctx)
+#ifndef NDEBUG
+		hal_consolePrint(ATTR_BOLD, "EFAULT in _threads_checkSignal, killing whole process\n");
+#endif
+		hal_spinlockClear(&threads_common.spinlock, &sc);
```
spinlock double unlock on line 1525

etiaro committed:
WIP: enable nightly tests in CI

etiaro commented on hal/ia32/cpu.c:
```diff
@@ -186,13 +188,21 @@ int hal_cpuPushSignal(void *kstack, void (*handler)(void), cpu_context_t *signal
-	signalCtx->eip = (u32)handler;
-	signalCtx->esp -= sizeof(cpu_context_t);
+	if (!hal_setexcjmp(&excctx, &oldctx)) {
+		hal_memcpy(signalCtx, ctx, sizeof(cpu_context_t));
```
Additional check for access to kernel memory map should be added

etiaro committed:
vm/map: mapBelongs protection check

Introduce checks for read/write access to memory pointed to in syscall
arguments.

TASK: RTOS-1141

etiaro committed:
syscalls: CPU fault handling

Protect userpace memory accesses with CPU fault handling mechanisms to
avoid TOCTOU errors.

TASK: RTOS-1141

ziemleszcz commented on syscalls.c:
```diff
@@ -915,24 +916,22 @@ int syscalls_msgRespond(u8 *ustack)
 		return -EFAULT;
 	}
 
-#ifndef NOMMU /* o.data has client memory pointer on NOMMU */
```
What about `NOMMU_MAPBELONGS_VALIDATE_MAP` that you have added earlier?

etiaro replied:
This ifndef seems outdated to me. 
With that flag, the check is still reasonable (if the process has write access to the output buffer it will return EOK, if the process does not have the access, the output buffer would be copied into his memory in msg-nommu.c).

Before, or without that flag, mapBelongs implementation was disabled on NOMMU for this very reason, so this seems like a double-disable.

ziemleszcz commented on hal/armv7a/exceptions.c:
```diff
@@ -118,12 +120,46 @@ static void exceptions_defaultHandler(unsigned int n, exc_context_t *ctx)
 /* parasoft-suppress-next-line MISRAC2012-RULE_8_4 "Usage in assembly" */
 void exceptions_dispatch(unsigned int n, exc_context_t *ctx)
 {
+	excjmp_context_t *excctx = threads_getexcjmp();
```
Shouldn't this be called from `map_pageFault()` on `vm_mapForce()` failure for USERMEM_TRY section?

etiaro replied:
Yeah, you're right. Previous solution would break on lazy allocation (if it would be enabled), now it should be fine, however exception dump before those checks from the master is kept.

etiaro committed:
vm/map: minimal mapBelongs with excjmp supported

With excjmp support all syscall arguments access should be protected via
kernel exception handling which allows to reduce mapbelongs to userspace
checking if address fits in userspace map range, reducing its overhead.

TASK: RTOS-1141

etiaro committed:
proc/threads: fault handling in threads_canaryInit

TASK: RTOS-1141

etiaro committed:
syscalls: handle faults on GETFROMSTACK macros

TASK: YT-1141
